### PR TITLE
Add default ctor, copy ctor and dtor for multiple CPP bindings

### LINKED ISSUE #753: Bindings for `std::unordered_map` are wrong
```c++
#include <unordered_map>
char (*__kaboom)[sizeof(std::unordered_map<int, int>)] = 1;
```
(Trick taken from https://stackoverflow.com/a/35261673/1021344)

This will print:
```console
clang++ uno.cpp
uno.cpp:3:8: error: cannot initialize a variable of type 'char (*)[40]' with an rvalue of type 'int'
char (*__kaboom)[sizeof(std::unordered_map<int, int>)] = 1;
       ^                                                 ~
1 error generated.
```

So the size is 40. However our bindings only have [*a single pointer*](https://github.com/bpfkorea/agora/blob/eb4ab3b0083c72fa2b7eb815ad2cd7a7290ffad4/source/scpd/Cpp.d#L85-L107).

That's on OSX with the latest clang++. I expect things to be wrong on other platforms too.

### LINKED ISSUE #2015: Switch SCP build from g++ to clang++
https://github.com/bosagora/agora/blob/c1d67bf97c663e2b4e6c2db118f693aef034283e/source/scpp/build.d#L62

We should be using `clang++` here. We are doing so already on Mac, because `g++` just aliases to `clang++`, but not yet on Linux.

### PR DESCRIPTION
**Pull request by omerfirmak**
Fixes #753 
Fixes #2015

### CODE REVIEW

File: source/scpd/Cpp.d [hunk 1]
```diff
@@ -76,40 +79,39 @@ private nothrow @nogc extern(C++) void cpp_unordered_map_assign (K, V)(
...
-
-/// Create a new unordered map with the given Key/T types, and return a pointer to it
-private pure nothrow @nogc @safe extern(C++) void* cpp_unordered_map_create (K, V)();
-
 /// Rudimentary bindings for std::unordered_map
 extern(C++, (StdNamespace))
 public struct unordered_map (Key, T, Hash = hash!Key, KeyEqual = equal_to!Key, Allocator = allocator!(pair!(const Key, T)))
 {
-    private void* ptr;
+    private ulong[56 / ulong.sizeof] _data;
```
Comment (Geod24) [inline source/scpd/Cpp.d]: `ulong.sizeof` is *always* 8. Did you mean `size_t` or `cpp_ulong` ?

Comment (AndrejMitrovic) [inline source/scpd/Cpp.d]: How'd you get this number?

Comment (omerfirmak) [reply source/scpd/Cpp.d]: It's a magic number, verified here

https://github.com/bosagora/agora/pull/2066/files#diff-2fff5f741967021f44722f88e37effa07123b287d13e843ff246b43d91a7d8d4R478

Comment (omerfirmak) [reply source/scpd/Cpp.d]: Nah, I made that a `ulong[]` hoping that it would help with alignment and hard coding length to 7 just looked odd.

Comment (omerfirmak) [reply source/scpd/Cpp.d]: It will probably be different for Mac OSX tho, CI will tell

Comment (AndrejMitrovic) [reply source/scpd/Cpp.d]: Yes I'm having similar problems with https://github.com/bosagora/agora/pull/2065. There's a size mismatch on one platform (linux) but not another (macos). So much for C++ portability huh? :o)

Comment (omerfirmak) [reply source/scpd/Cpp.d]: Our `shared_ptr`binding is faulty too, it causes C++ runtime to free the pointer we pass to it. Maybe this and https://github.com/bosagora/agora/pull/2016 could help you.

File: source/scpd/Cpp.d [hunk 2]
```diff
@@ -76,40 +79,39 @@ private nothrow @nogc extern(C++) void cpp_unordered_map_assign (K, V)(
...
+
+    @disable this ();
+
+    extern(D):
+    this (char dummy)
+    {
+        defaultCtorCPPObject!(typeof(this))(&this);
+    }
```
Comment (Geod24) [inline source/scpd/Cpp.d]: What's the reasoning for having a dummy ctor over, say, a `static` `make` function ?

File: source/scpp/extra/DUtils.h [hunk 3]
```diff
@@ -74,3 +74,21 @@ VectorT duplicate(const VectorT& this_)
...
+void defaultCtorCPPObject (T* ptr)
+{
+    new(ptr) T();
+}
+
+template<typename T>
+void dtorCPPObject (T* ptr)
+{
+    ptr->~T();
+}
```
Comment (Geod24) [inline source/scpp/extra/DUtils.h:94]: Could https://github.com/dlang/druntime/blob/661c2d8975d5a5c6bfbab6b18850059aac857046/src/object.d#L3955-L3971 help ?

File: source/scpd/quorum/QuorumTracker.d [hunk 4]
```diff
@@ -70,6 +70,3 @@ public:
     /// returns the known quorum map of the entire network
     ref const(QuorumMap) getQuorum () const;
 }
-
-// should be 48, see #753
-static assert(QuorumTracker.sizeof == 16);
```
Comment (Geod24) [inline source/scpd/quorum/QuorumTracker.d:75]: Why is this removed and not fixed ?

File: source/scpd/Cpp.d (hunk 2 above)
Comment (omerfirmak) [reply source/scpd/Cpp.d]: I think this lets you construct the object anywhere (stack/heap), without explicitly passing a pointer or copying the whole thing, right?

File: source/scpd/quorum/QuorumTracker.d (hunk 4 above)
Comment (omerfirmak) [reply source/scpd/quorum/QuorumTracker.d:75]: While I think that does not really serve any purpose, I'll try to rewrite it better.

File: source/scpp/extra/DUtils.h (hunk 3 above)
Comment (omerfirmak) [reply source/scpp/extra/DUtils.h:94]: I'll give it a try but in the cases where linker can't link `~object()` directly, that probably won't work either.

File: source/scpp/extra/DUtils.h [hunk 5]
```diff
@@ -74,3 +74,21 @@ VectorT duplicate(const VectorT& this_)
...
+template<typename T>
+int getCPPSizeof ()
+{
+    return sizeof(T);
+}
+
+template<typename T>
+void defaultCtorCPPObject (T* ptr)
+{
+    new(ptr) T();
```
Comment (linked0) [inline source/scpp/extra/DUtils.h:87]: Is it the same as follows?
`new T(ptr);`

Comment (linked0) [reply source/scpp/extra/DUtils.h:87]: Ah, there was some misunderstanding. Never mind. :)

**Comment by omerfirmak** [comment]
All green

**Comment by omerfirmak** [comment]
Apart from `std::vector`, all containers now have proper copy ctor and `opAssign`

File: source/scpd/Cpp.d [hunk 6]
```diff
@@ -34,7 +34,72 @@ import std.meta;
...
+    void copyCtorCPPObject(T) (T* ptr, inout(T)* rhs);
+    void copyCtorCPPObject(T) (immutable(T)* ptr, inout(T)* rhs);
+    int getCPPSizeof(T) ();
+}
+
+private mixin template CPPBindingMixin (T)
+{
+    extern(D)
+    {
+        this (char dummy) @trusted nothrow @nogc
```
Comment (AndrejMitrovic) [inline source/scpd/Cpp.d]: Would it be a good idea to add `@disable this();` so people remember to call this ctor?

File: source/scpd/types/XDRBase.d [hunk 7]
```diff
@@ -45,48 +45,32 @@ uint size32(size_t s)
 ////////////////////////////////////////////////////////////////
 
 //! Generic class of XDR unmarshaling errors.
-class xdr_runtime_error : runtime_error
-{
-}
```
Comment (Geod24) [inline source/scpd/types/XDRBase.d]: Why ? `alias` is almost like a `#define` for type, so it would e.g. give the wrong name if used as template parameter.

File: source/scpd/Cpp.d (hunk 6 above)
Comment (Geod24) [reply source/scpd/Cpp.d]: Yes to `@disable this();`

I would also suggest creating a special enum and only using this, e.g. `enum CppCtor { Use = 0; }` and then:
```D
public this (CppCtor use);
auto x = CppType(CppCtor.Use);
```

**Review by Geod24 — COMMENTED** [review]
How does the second commit work given we can't pass `CppRuntime_Clang` to LDC ?

File: source/scpd/types/XDRBase.d (hunk 7 above)
Comment (omerfirmak) [reply source/scpd/types/XDRBase.d]: Reverted those

File: source/scpd/Cpp.d (hunk 6 above)
Comment (omerfirmak) [reply source/scpd/Cpp.d]: Yep, default ctor was meant to be disabled.

**Comment by omerfirmak** [comment]
> How does the second commit work given we can't pass `CppRuntime_Clang` to LDC ?

Well, it just changes the compiler but we still use `libstd++` on linux.

File: source/scpd/types/XDRBase.d (hunk 7 above)
Comment (omerfirmak) [reply source/scpd/types/XDRBase.d]: and it immediately broke macos.

**Comment by Geod24** [comment]
```
Undefined symbols for architecture x86_64:
  "std::__1::logic_error::what() const", referenced from:
      __D4scpd5types7XDRBase15xdr_wrong_union6__vtblZ in agora.o
      __D4scpd3Cpp11logic_error6__vtblZ in agora.o
  "std::__1::runtime_error::what() const", referenced from:
      __D4scpd5types7XDRBase17xdr_runtime_error6__vtblZ in agora.o
      __D4scpd5types7XDRBase12xdr_overflow6__vtblZ in agora.o
      __D4scpd5types7XDRBase18xdr_stack_overflow6__vtblZ in agora.o
      __D4scpd5types7XDRBase20xdr_bad_message_size6__vtblZ in agora.o
      __D4scpd5types7XDRBase20xdr_bad_discriminant6__vtblZ in agora.o
      __D4scpd5types7XDRBase18xdr_should_be_zero6__vtblZ in agora.o
      __D4scpd5types7XDRBase20xdr_invariant_failed6__vtblZ in agora.o
```

Don't think Exceptions are in `__1` are they ?

**Comment by Geod24** [comment]
```
[source/agora/test/Flash.d:1343] Assertion thrown during test: assert(info.state == ChannelState.SettingUp || info.state == ChannelState.WaitingForFunding) failed
/Users/runner/work/agora/agora/source/agora/test/Flash.d:1343 void agora.test.Flash.__unittest_L1211_C1() [0x11006133e]
/Users/runner/work/agora/agora/source/agora/test/Flash.d agora.test.Flash.__unittest [0x110079c5e]
/Users/runner/work/agora/agora/source/agora/test/Base.d:264 bool agora.test.Base.customModuleUnitTester().runTest(agora.test.Base.customModuleUnitTester().ModTest) [0x110099a60]
/Users/runner/work/agora/agora/source/agora/test/Base.d:283 core.runtime.UnitTestResult agora.test.Base.customModuleUnitTester() [0x11009835c]
runtime.d:607 runModuleUnitTests [0x111109a28]
dmain2.d:551 void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x1111236e2]
dmain2.d:525 void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x111123662]
dmain2.d:584 _d_run_main2 [0x11112351d]
dmain2.d:384 _d_run_main [0x1111231d4]
/Users/runner/hostedtoolcache/dc/ldc2-1.26.0/x64/ldc2-1.26.0-osx-x86_64/bin/../import/core/internal/entrypoint.d:42 main [0x1100a4774]
```

Never seen this one before

**Comment by omerfirmak** [comment]
green after multiple tries

**Comment by omerfirmak** [comment]
Needs approval too @Geod24

**Review by Geod24 — APPROVED** [review]
